Add traffic mirroring for Istio service mesh

andrewjjenkins opened the pull request:
Traffic mirroring is a pre-stage for canary deployments.  When mirroring
is enabled, at the beginning of a canary deployment traffic is mirrored
to the canary instead of shifted for one canary period.  The service
mesh should mirror by copying the request and sending one copy to the
primary and one copy to the canary; only the response from the primary
is sent to the user.  The response from the canary is only used for
collecting metrics.

Once the mirror period is over, the canary proceeds as usual, shifting
traffic from primary to canary until complete.

Traffic mirroring can also be used for Blue/Green deployments Fix: #313

andrewjjenkins commented:
I marked this as work-in-progress as I think it needs some high-level review and deciding what adjustments need to be made to the approach.

This implementation creates a traffic mirroring pre-stage for canary deployments.  It does this by adding a field to the canary spec called [Mirror](https://github.com/weaveworks/flagger/pull/311/files#diff-f08875c940c9f976c794c6da8180564d).  If true, we run a pre-stage where we tell Istio to mirror traffic to the canary.  Here's an example prometheus plot:

![image](https://user-images.githubusercontent.com/803647/65461860-d9b77500-de11-11e9-88eb-3398e33baa45.png)

The top teal line is primary, the red line is mirrored traffic to canary, the green line is shifted traffic to canary.

First question I have is what should be the appropriate behavior if the service mesh doesn't support mirroring?  I implemented and tested for Istio but some meshes & SMI don't support mirroring.

andrewjjenkins commented on pkg/controller/scheduler.go:
```diff
@@ -446,13 +447,34 @@ func (c *Controller) advanceCanary(name string, namespace string, skipLivenessCh
+			// FIXME(andrewj): Is this a bug?  Should floor(primaryWeight, 100)?
+			if primaryWeight > 100 {
+				primaryWeight = 100
+			}
```
Wasn't sure about this, I think this maybe should have always been:

```
if canaryWeight > 100 {
  canaryWeight = 100
}
```

stefanprodan replied:
Good catch! canary weight should be `100 - primaryWeight`

andrewjjenkins replied:
Already fixed upstream after rebase.

stefanprodan commented on pkg/controller/scheduler.go:
```diff
@@ -446,13 +447,34 @@ func (c *Controller) advanceCanary(name string, namespace string, skipLivenessCh
+				canaryWeight = cd.Spec.CanaryAnalysis.StepWeight
+			}
+			c.logger.With("canary", fmt.Sprintf("%s.%s", name, namespace)).
+				Warnf("Running mirror step %d/%d/%t", primaryWeight, canaryWeight, mirrored)
```
I think this should be `Infof` instead of `Warnf `.

andrewjjenkins replied:
Yes sorry I think there are a few log messages that may be extraneous or at too high of a level.  Let me do a pass for all logs I added.

stefanprodan commented:
@andrewjjenkins I think mirroring should be enabled for Blue/Green deployments as well.

>First question I have is what should be the appropriate behavior if the service mesh doesn't support mirroring? I implemented and tested for Istio but some meshes & SMI don't support mirroring.

For SMI and others, mirroring should be off be default, the code looks ok as it this. Thanks

stefanprodan commented:
I have changed the way promotion works in #310 I'll merge it soon, please rebase with master after the merge, as some changes could conflict with your branch.

andrewjjenkins commented:
OK, rebased in #310, reduced the log message you pointed out to Infof and deleted some others that were left over from debugging.

Still remaining: Blue/Green implementation and I think I should add this to the docs as well.

andrewjjenkins commented:
@stefanprodan I think I've addressed your comments & work that I'm aware of.
- Rebased on #310 
- Fixed log messages, set one to Infof and removed the rest.
- Added Blue/Green implementation
- Added to docs

The B/G implementation: if "mirror: true" then mirror traffic for the entire canary analysis phase.  Open to suggestions here if you had a different idea on how it should work.

Thanks & I think it's ready for you to take a look.

andrewjjenkins commented on pkg/controller/scheduler.go:
```diff
@@ -372,6 +373,15 @@ func (c *Controller) advanceCanary(name string, namespace string, skipLivenessCh
 		// increment iterations
 		if cd.Spec.CanaryAnalysis.Iterations > cd.Status.Iterations {
+			// If in "mirror" mode, mirror requests during the entire B/G canary test
+			if provider != "kubernetes" &&
```
I'm only turning mirroring on if `provider != kubernetes` because down below we only route traffic to the canary (turn mirroring off) if `provider != kubernetes`.

I think this is fine and doesn't matter because `provider == kubernetes` is probably never going to work since you can't clone requests AFAIK.  But heads up, please LMK if there's a better way to do this.

stefanprodan replied:
This is great, no need to call the mesh router if Kubernetes L4 is in use 👍

andrewjjenkins commented:
Sorry for the delay, added to those three copies of the CRD and rebased in (see e.g. https://github.com/weaveworks/flagger/pull/311/commits/e384b03d49fb43bc4858d7b2d5404dfba37df92d#diff-5614edd054a6dc84c5fab6ccd18b9600 ).

andrewjjenkins commented:
Thanks, I don't see a merge button so I'll wait for you to click it.
